components/net/filemanager_thread.rs

```diff
@@ -108,109 +109,205 @@ enum FileImpl {
+            }
+            Err(e) => {
+                send_error(load_data.url.clone(), NetworkError::Internal(format!("{:?}", e)), consumer);
+            }
+        }
+    }
+}
+
+struct FileManagerStore<UI: 'static + UIProvider> {
+    entries: RwLock<HashMap<Uuid, FileStoreEntry>>,
+    ui: &'static UI,
+}
+
+impl <UI: 'static + UIProvider> FileManagerStore<UI> {
+    fn new(ui: &'static UI) -> Self {
+        FileManagerStore {
+            entries: RwLock::new(HashMap::new()),
+            ui: ui,
+        }
+    }
+
+    /// Copy out the file backend implementation content
+    fn get_impl(&self, id: &Uuid, origin_in: &FileOrigin) -> Result<FileImpl, BlobURLStoreError> {
+        match self.entries.read().unwrap().get(id) {
+            Some(ref e) => {
+                if *origin_in != *e.origin {
+                    Err(BlobURLStoreError::InvalidOrigin)
+                } else {
+                    Ok(e.file_impl.clone())
+                }
+            }
+            None => Err(BlobURLStoreError::InvalidFileID),
+        }
+    }
+
+    fn insert(&self, id: Uuid, entry: FileStoreEntry) {
+        self.entries.write().unwrap().insert(id, entry);
+    }
+
+    fn remove(&self, id: &Uuid) {
+        self.entries.write().unwrap().remove(id);
+    }
+
+    fn inc_ref(&self, id: &Uuid, origin_in: &FileOrigin) -> Result<(), BlobURLStoreError>{
+        match self.entries.read().unwrap().get(id) {
             Some(entry) => {
-                if entry.origin == origin_in {
-                    entry.refs.set(entry.refs.get() + 1);
+                if entry.origin == *origin_in {
+                    entry.refs.fetch_add(1, Ordering::Relaxed);
+                    Ok(())
+                } else {
+                    Err(BlobURLStoreError::InvalidOrigin)
                 }
             }
-            None => return, // Invalid UUID
+            None => Err(BlobURLStoreError::InvalidFileID),
         }
     }
 
-    fn add_sliced_entry(&mut self, id: SelectedFileId, rel_pos: RelativePos,
+    fn add_sliced_entry(&self, parent_id: SelectedFileId, rel_pos: RelativePos,
                         sender: IpcSender<Result<SelectedFileId, BlobURLStoreError>>,
                         origin_in: FileOrigin) {
-        if let Ok(id) = Uuid::parse_str(&id.0) {
-            match self.store.get(&id) {
-                Some(entry) => {
-                    if entry.origin == origin_in {
-                        // inc_ref on parent entry
-                        entry.refs.set(entry.refs.get() + 1);
-                    } else {
-                        let _ = sender.send(Err(BlobURLStoreError::InvalidOrigin));
-                        return;
-                    }
-                },
-                None => {
-                    let _ = sender.send(Err(BlobURLStoreError::InvalidFileID));
-                    return;
+        if let Ok(parent_id) = Uuid::parse_str(&parent_id.0) {
+            match self.inc_ref(&parent_id, &origin_in) {
```

Comment:
It might be better if the incref and insert use the same lock. Technically still safe if we don't (we need to be careful about this for decrefs though).